04-Protocol-Security.md

```diff
@@ -0,0 +1,407 @@
+3. calls `MixHash(e.public_key)`
+4. calls `EncryptAndHash()` with empty payload and appends the ciphertext to the buffer (note that *k* is empty at this point, so this effectively reduces down to `MixHash()` on empty data)
+5. submits the buffer for sending to the responder in the following format
+
+```
+Ephemeral public key message:
++---------------+------------------------------------------------------------------------------------------------------+
+| PUBKEY        |  Initiator's ephemeral public key                                                                    |
++---------------+------------------------------------------------------------------------------------------------------+
+
+Message length: 32 bytes
+```
+
+#### 4.5.1.2 Responder
+1. receives ephemeral public key message (32 bytes plaintext public key)
+2. parses received public key as `re.public_key`
+3. calls `MixHash(re.public_key)`
+4. calls `DecryptAndHash()` on remaining bytes (i.e. on empty data with empty *k*, thus effectively only calls `MixHash()` on empty data)
+
+### 4.5.2 Handshake Act 2: NX-handshake part 2 `<- e, ee, s, es, SIGNATURE_NOISE_MESSAGE`
+
+Responder provides its ephemeral, encrypted static public keys and encrypted `SIGNATURE_NOISE_MESSAGE` to the initiator, performs Elliptic-Curve Diffie-Hellman operations.
+
+```
+SIGNATURE_NOISE_MESSAGE
++-----------------+-----------+----------------------------------------------------------------------------------------+
+| Field Name      | Data Type | Description                                                                            |
++-----------------+-----------+----------------------------------------------------------------------------------------+
+| version         | U16       | Version of the certificate format                                                      |
++-----------------+-----------+----------------------------------------------------------------------------------------+
+| valid_from      | U32       | Validity start time (unix timestamp)                                                   |
++-----------------+-----------+----------------------------------------------------------------------------------------+
+| not_valid_after | U32       | Signature is invalid after this point in time (unix timestamp)                         |
++-----------------+-----------+----------------------------------------------------------------------------------------+
+| signature       | SIGNATURE | Certificate signature                                                                  |
++-----------------+-----------+----------------------------------------------------------------------------------------+
+
+Length: 74 bytes
+```
+
+
+#### 4.5.2.1 Responder
+1. initializes empty output buffer
+2. generates ephemeral keypair `e`, appends `e.public_key` to the buffer (32 bytes plaintext public key)
+3. calls `MixHash(e.public_key)`
+4. calls `MixKey(ECDH(e, re))`
```

Comment:
Perhaps is ECDH(e.priv, re)? The same for ECDH(s.priv, re) at line 225

Reply:
Is that really necessary? Because the case is obvious. ECDH is performed between local party's private key and remote party's public key.

Reply:
to write ECDH(e, re) is slightly wrong because e is a keypair in a place where there should be a private key only, but I think that almost everyone that reads it will understand if you leave as it is.

Reply:
ok, so let's write it explicitly.